### PR TITLE
executor: fix query hang in IndexMerge Executor when it's killed

### LINKED ISSUE #45279: The query may hang in IndexMerge executor when it's killed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. 
```sql
CREATE TABLE reproduce (
    c1 int primary key,
    c2 int,
    c3 int,
    key ci2(c2),
    key ci3(c3)
);
```

2. insert lots of rows about 100 million
3. select * from reproduce where c1 in (0, 10, 20, 30, ...., 10000) or c2 in (0, 10, 20, 30, ...., 10000) // make this query run for a long time
4. kill the query executed in step 3

### 2. What did you expect to see? (Required)

Query should be successfully killed

### 3. What did you see instead (Required)

Query hang can not be successfully killed.

### 4. What is your TiDB version? (Required)

Commit Hash: dd60929c5

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #45279 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/index_merge_reader.go [hunk 1]
```diff
@@ -807,7 +807,9 @@ func (e *IndexMergeReaderExecutor) Next(ctx context.Context, req *chunk.Chunk) e
 
 	req.Reset()
 	for {
-		resultTask, err := e.getResultTask()
+		ctx1, cancel := context.WithCancel(ctx)
```
Comment (guo-shaoge) [inline executor/index_merge_reader.go:810]: Why not use `ctx` directly?

Comment (xzhangxian1008) [reply executor/index_merge_reader.go]: > Why not use `ctx` directly?

done

File: executor/index_merge_reader.go [hunk 2]
```diff
@@ -839,8 +839,14 @@ func (e *IndexMergeReaderExecutor) getResultTask() (*indexMergeTableTask, error)
 	if !ok {
 		return nil, nil
 	}
-	if err := <-task.doneCh; err != nil {
-		return nil, errors.Trace(err)
+
+	select {
+	case <-ctx.Done():
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply executor/index_merge_reader.go:845]: > Do we need to return ctx.Err()

okk, I will return ctx.Err()

**Review by guo-shaoge — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**